backend/cmm_helpers.ml

```diff
@@ -2059,18 +2067,50 @@ let ptr_offset ptr offset dbg =
 let direct_apply lbl args (pos, _mode) dbg =
   Cop (Capply (typ_val, pos), Cconst_symbol (lbl, dbg) :: args, dbg)
 
+let call_caml_apply mut clos args pos mode dbg =
+  let arity = List.length args in
+  let really_call_caml_apply clos args =
+    let cargs =
+      (Cconst_symbol (apply_function_sym arity mode, dbg) :: args) @ [clos]
+    in
+    Cop (Capply (typ_val, pos), cargs, dbg)
+  in
+  if !Flambda_backend_flags.caml_apply_inline_fast_path
+  then
+    (*    (if (= clos.arity N)
+     *      (app clos.direct a1 ... aN clos)
+     *      (app call_caml_applyN a1 .. aN clos)
+     *)
+    bind_list "arg" args (fun args ->
+        bind "fun" clos (fun clos ->
+            Cifthenelse
+              ( Cop
+                  ( Ccmpi Ceq,
+                    [ Cop
+                        ( Casr,
+                          [ get_field_gen mut clos 1 dbg;
```

Comment:
(I am not sure I understand why this call,
and the one below on line 2099, are not
using `Immutable` for these two fields.)

Reply:
Because some closures are allocated and modified in place (those bound by a let rec with non-functional values mixed in). Something related has been discussed upstream recently: ocaml/ocaml#11654